### DISCUSSION TITLE
Equal-weighted portfolios transaction costs

### DISCUSSION THREAD

**Discussion by datascienceknowledgea** [skfolio/skfolio#43] [General]
Hi,
how is it possible to take transaction costs in a multiperiod equal-weighted portfolio into account?

**Comment by HugoDelatte** [comment]
You can follow [this example](https://skfolio.org/auto_examples/1_mean_risk/plot_6_transaction_costs.html#sphx-glr-auto-examples-1-mean-risk-plot-6-transaction-costs-py) that builds a multiperiod `MeanRisk` Portfolio. The only difference is that if you use an estimator that is not impacted by the transaction costs (which is the case of `EqualWeighted` as the weights will always be 1/N regardless of TC), you need to pass the `transaction_costs` and `previous_weights` to the Portoflio using the param `portfolio_params` of the `EqualWeighted`  estimator (details in the doctsrings).

**Reply by datascienceknowledgea** [reply]
Do you mean like this?
When comparing the returns before and after transaction costs it just subtracts 0.002/holding_period from every return. 
```
holding_period = 100
fitting_period = 1000
cv = WalkForward(train_size=fitting_period, test_size=holding_period)

transaction_costs = 0.002 / holding_period
previous_weights = np.ones(3) / 3



model = EqualWeighted(portfolio_params={'transaction_costs':transaction_costs, 'previous_weights':previous_weights})

# pred 11 as Multiperiodportfolio
pred11 = cross_val_predict(model, X, cv=cv, n_jobs=-1)
pred12 = MultiPeriodPortfolio()
previous_weights = None
for portfolio in pred11:
    new_portfolio = Portfolio(
        X=portfolio.X,
        weights=portfolio.weights,
        previous_weights=previous_weights,
        transaction_costs=transaction_costs,
    )
    previous_weights = portfolio.weights
    pred12.append(new_portfolio)
```

**Reply by HugoDelatte** [reply]
Yes that's correct:

![image](https://github.com/skfolio/skfolio/assets/50495273/c67dee8e-3e60-4869-b7f2-66d8d21e578d)

For a `MultiPeriodPortfolio` based on Equal Weighted Portfolios, `w_i` is always equal to `w_prev_i` at each rebalancing except for the first one because you passed `None` for the initial Portfolio.

**Reply by datascienceknowledgea** [reply]
Thank you very much. I would like to take transaction costs into account at each rebalancing, but this isn't possible because w_i = w_prev_i. How could this be solved? Have you thought about implementing a weight drift instead of assuming constant weights?

**Reply by HugoDelatte** [reply]
I guess that what you are saying is that in reality, because of the weight drift (due to spot move), your weights at the end of the period are not equi-weighted anymore and therefore you have some TC when rebalancing to an equal-weighted allocation.

Modeling weight drift is a task for a backtesting package. Cross-validators in skfolio (as well as scikit-learn) are used for model validation, selection and parameter tuning. Therefore, the model evaluation needs to be homogenous to the optimization problems. For example, if you call `model.fit_predict(X)` with model a `MeanRisk` with volatility constraint of 10%, your predicted portfolio on the same training set `X` must have a volatility of 10%, otherwise you start creating a lot of problems. Mathematically, the portfolio returns need to be computed using dot product: $R \cdot w-costs$

Related to [this discussion](https://github.com/skfolio/skfolio/discussions/41#discussioncomment-8725239)

**Reply by CarloNicolini** [reply]
@HugoDelatte this reminds me about our discussion on weight drift. The user @datascienceknowledgea just opened an issue on scikit-portfolio about backtesting where weight drift is actually implemented by simply rolling over each day and recomputing actual positions and every "backtesting period" , rebalance the weights.

Hugo, what do you think about the possibility of joining forces for a new skfolio based package performing vectorial backtesting that supports weights drift? Alternatively, add another kind of `predict_drift` method supporting this idea within the portfolio estimator.

**Reply by HugoDelatte** [reply]
Yes let's explore existing backtesting libraries and see where we can add value. There are already a lot of python backtesters so I always thought that innovation was limited , but actually we could potentially add value on the API design, creating a backtester closer to skfolio and scikit-learn API.
